Failed linking due to -lomp in DEFAULT_SEQUENTIAL_LINK_LINE

**Bug summary**

When libomp.so is located in a non-standard path, `syclcc` will fail at linking when targeting a GPU due to `default-sequential-link-line` having `-lomp` but no path to it.

**To Reproduce**

Have clang installed in a non-default location. Build hipSYCL with a GPU backend. Then, try to compile a simple program:

Sample dockerfile:

```dockerfile
# Use CUDA image as base
FROM nvidia/cuda:11.4.3-devel-ubuntu20.04

# Standard stuff
RUN apt-get update -y && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        apt-transport-https \
        build-essential \
        ca-certificates \
        cmake \
        git \
        gnupg \
        gpg-agent \
        libboost-all-dev \
        moreutils \
        vim \
        wget \
        && \
    rm -rf /var/lib/apt/lists/*

# LLVM-14
RUN wget -qO - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - && \
    echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main" >> /etc/apt/sources.list.d/hpccm.list && \
    echo "deb-src http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main" >> /etc/apt/sources.list.d/hpccm.list && \
    apt-get update -y && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        clang-14 \
        libomp-14-dev \
        libclang-14-dev \
        libomp5-14 \
        lld-14 \
        llvm-14-dev \
        && \
    rm -rf /var/lib/apt/lists/*
RUN update-alternatives --install /usr/bin/clang clang $(which clang-14) 30 && \
    update-alternatives --install /usr/bin/clang++ clang++ $(which clang++-14) 30

# hipSYCL
RUN mkdir -p /var/tmp && cd /var/tmp && git clone --recursive https://github.com/illuhad/hipSYCL.git hipSYCL && cd - && cd /var/tmp/hipSYCL && git checkout c828db0 && cd - && \
    mkdir -p /var/tmp/hipSYCL/build && cd /var/tmp/hipSYCL/build && cmake -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang-14 -DCMAKE_CXX_COMPILER=clang++-14 -DLLVM_DIR=/usr/lib/llvm-14/cmake/ -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda -DWITH_CUDA_BACKEND=ON /var/tmp/hipSYCL && \
    cmake --build /var/tmp/hipSYCL/build --target install -- -j$(nproc) && \
    rm -rf /var/tmp/hipSYCL

# Trigger the bug
RUN echo "int main() {}" > /test.cpp && syclcc --hipsycl-targets="cuda:sm_75" /test.cpp
```

Observed behavior:

```
/usr/bin/ld: cannot find -lomp
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Expected behavior**

No error.

**Describe your setup**
* c828db0
* Describe the dependencies that hipSYCL sits on top of in your setup: 
   * clang-14 from LLVM APT repos
   * CUDA 11.4.3 from NVIDIA docker image

**Optional additional diagnostic information**

Either of the following resolves the problem:

- Add `omp` to `--hipsycl-targets`.
- Add `-L/usr/lib/llvm-14/lib/` to `default-sequential-link-line`.
- Remove `-lomp` from `default-sequential-link-line`.

**Proposed solution**

In https://github.com/illuhad/hipSYCL/blob/ce37cf8d35be120fc066c792b89ac82cacb71eaa/CMakeLists.txt#L248, we shall either remove `-lomp` (if we genuinely don't need OpenMP) or use `${DEFAULT_OMP_FLAG}` if we need it.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -245,7 +245,7 @@ set(DEFAULT_WIN32_SEQUENTIAL_LINK_LINE "-L${Boost_LIBRARY_DIR} ${DEFAULT_LINK_PR
 set(DEFAULT_APPLE_OMP_LINK_LINE "${DEFAULT_BOOST_LIBRARIES} ${DEFAULT_OMP_FLAG} ${hipSYCL_OpenMP_CXX_LIBRARIES}")
 set(DEFAULT_APPLE_SEQUENTIAL_LINK_LINE "${DEFAULT_BOOST_LIBRARIES} ${hipSYCL_OpenMP_CXX_LIBRARIES}")
 set(DEFAULT_OMP_LINK_LINE "-L${Boost_LIBRARY_DIR} -lboost_context -lboost_fiber -Wl,-rpath=${Boost_LIBRARY_DIR} ${DEFAULT_OMP_FLAG}")
-set(DEFAULT_SEQUENTIAL_LINK_LINE "-L${Boost_LIBRARY_DIR} -lboost_context -lboost_fiber -lomp -Wl,-rpath=${Boost_LIBRARY_DIR}")
+set(DEFAULT_SEQUENTIAL_LINK_LINE "-L${Boost_LIBRARY_DIR} -lboost_context -lboost_fiber -Wl,-rpath=${Boost_LIBRARY_DIR}")
 
 # If no link lines given, set to default.
 if(NOT ROCM_LINK_LINE)

--- a/include/hipSYCL/glue/generic/host/range_decomposition.hpp
+++ b/include/hipSYCL/glue/generic/host/range_decomposition.hpp
@@ -29,7 +29,6 @@
 #ifndef HIPSYCL_RANGE_DECOMPOSITION_HPP
 #define HIPSYCL_RANGE_DECOMPOSITION_HPP
 
-#include <omp.h>
 #include <vector>
 #include <cassert>
 

--- a/include/hipSYCL/glue/omp/omp_kernel_launcher.hpp
+++ b/include/hipSYCL/glue/omp/omp_kernel_launcher.hpp
@@ -30,7 +30,9 @@
 
 #include <cassert>
 #include <tuple>
+#ifdef _OPENMP
 #include <omp.h>
+#endif
 
 #include "hipSYCL/common/debug.hpp"
 #include "hipSYCL/runtime/error.hpp"
@@ -62,10 +64,34 @@ namespace hipsycl {
 namespace glue {
 namespace omp_dispatch {
 
+inline int get_my_thread_id() {
+#ifdef _OPENMP
+  return omp_get_thread_num();
+#else
+  return 0;
+#endif
+}
+
+inline int get_max_num_threads() {
+#ifdef _OPENMP
+  return omp_get_max_threads();
+#else
+  return 1;
+#endif
+}
+
+inline int get_num_threads() {
+#ifdef _OPENMP
+  return omp_get_num_threads();
+#else
+  return 1;
+#endif
+}
+
 template <class ReductionDescriptor> class omp_reducer {
 public:
   omp_reducer(host::sequential_reducer<ReductionDescriptor>& seq_reducer)
-      : _seq_reducer{seq_reducer}, _my_thread_id{omp_get_thread_num()} {}
+      : _seq_reducer{seq_reducer}, _my_thread_id{get_my_thread_id()} {}
 
   using value_type =
       typename host::sequential_reducer<ReductionDescriptor>::value_type;
@@ -90,12 +116,13 @@ void finalize_reduction(SequentialReducer &reducer) {
 template <class Function, typename... Reductions>
 void reducible_parallel_invocation(Function kernel,
                                    Reductions... reductions) noexcept {
-  int max_threads = omp_get_max_threads();
+  int max_threads = get_max_num_threads();
 
   auto sequential_reducers =
       std::make_tuple(host::sequential_reducer{max_threads, reductions}...);
-
+#ifdef _OPENMP
 #pragma omp parallel shared(sequential_reducers)
+#endif
   {
     auto make_omp_reducers = [&](auto &... seq_reducers) {
       return std::make_tuple(omp_reducer{seq_reducers}...);
@@ -120,19 +147,25 @@ template <int Dim, class Function>
 void iterate_range_omp_for(sycl::range<Dim> r, Function f) noexcept {
 
   if constexpr (Dim == 1) {
+#ifdef _OPENMP
     #pragma omp for
+#endif
     for (std::size_t i = 0; i < r.get(0); ++i) {
       f(sycl::id<Dim>{i});
     }
   } else if constexpr (Dim == 2) {
+#ifdef _OPENMP
     #pragma omp for collapse(2)
+#endif
     for (std::size_t i = 0; i < r.get(0); ++i) {
       for (std::size_t j = 0; j < r.get(1); ++j) {
         f(sycl::id<Dim>{i, j});
       }
     }
   } else if constexpr (Dim == 3) {
+#ifdef _OPENMP
     #pragma omp for collapse(3)
+#endif
     for (std::size_t i = 0; i < r.get(0); ++i) {
       for (std::size_t j = 0; j < r.get(1); ++j) {
         for (std::size_t k = 0; k < r.get(2); ++k) {
@@ -151,15 +184,18 @@ void iterate_range_omp_for(sycl::id<Dim> offset, sycl::range<Dim> r,
   const std::size_t max_i = offset.get(0) + r.get(0);
 
   if constexpr (Dim == 1) {
+#ifdef _OPENMP
   #pragma omp for
+#endif
     for (std::size_t i = min_i; i < max_i; ++i) {
       f(sycl::id<Dim>{i});
     }
   } else if constexpr (Dim == 2) {
     const std::size_t min_j = offset.get(1);
     const std::size_t max_j = offset.get(1) + r.get(1);
-
+#ifdef _OPENMP
   #pragma omp for collapse(2)
+#endif
     for (std::size_t i = min_i; i < max_i; ++i) {
       for (std::size_t j = min_j; j < max_j; ++j) {
         f(sycl::id<Dim>{i, j});
@@ -170,8 +206,9 @@ void iterate_range_omp_for(sycl::id<Dim> offset, sycl::range<Dim> r,
     const std::size_t min_k = offset.get(2);
     const std::size_t max_j = offset.get(1) + r.get(1);
     const std::size_t max_k = offset.get(2) + r.get(2);
-
+#ifdef _OPENMP
   #pragma omp for collapse(3)
+#endif
     for (std::size_t i = min_i; i < max_i; ++i) {
       for (std::size_t j = min_j; j < max_j; ++j) {
         for (std::size_t k = min_k; k < max_k; ++k) {
@@ -285,11 +322,11 @@ inline void parallel_for_ndrange_kernel(
     });
 #elif defined(HIPSYCL_HAS_FIBERS)
     host::static_range_decomposition<Dim> group_decomposition{
-        num_groups, omp_get_num_threads()};
+        num_groups, get_num_threads()};
 
     host::collective_execution_engine<Dim> engine{num_groups, local_size,
                                                   offset, group_decomposition,
-                                                  omp_get_thread_num()};
+                                                  get_my_thread_id()};
 
     std::function<void()> barrier_impl = [&]() { engine.barrier(); };
 

